src/parquet/reader-test.cc

```diff
@@ -149,13 +150,43 @@ TEST_F(TestAllTypesPlain, ColumnSelectionOutOfRange) {
 
   std::list<int> columns;
   columns.push_back(100);
-  ASSERT_THROW(reader_->DebugPrint(ss, columns), ParquetException);
+  ASSERT_THROW(reader_->DebugPrint(ss, columns, 128), ParquetException);
 
   columns.clear();
   columns.push_back(-1);
-  ASSERT_THROW(reader_->DebugPrint(ss, columns), ParquetException);
+  ASSERT_THROW(reader_->DebugPrint(ss, columns, 128), ParquetException);
 }
 
+TEST_F(TestAllTypesPlain, MemoryEstimation) {
+  std::stringstream ss;
+
+  int64_t batch_size = 128;
+  std::list<int> columns;
+
+  columns.push_back(0);
+  columns.push_back(5);
+  ParquetFileReader::MemoryUsage memory_usage1 =
+      reader_->EstimateMemoryUsage(false, columns, batch_size);
+  reader_->DebugPrint(ss, columns, batch_size, true);
+  ASSERT_GT(allocator_.MaxMemory(), memory_usage1.memory);
+  ASSERT_TRUE(memory_usage1.has_dictionary);
+
+  columns.push_back(9);
+  ParquetFileReader::MemoryUsage memory_usage2
+      = reader_->EstimateMemoryUsage(false, columns, batch_size);
+  reader_->DebugPrint(ss, columns, batch_size, true);
+  ASSERT_GT(allocator_.MaxMemory(), memory_usage2.memory);
+  ASSERT_TRUE(memory_usage2.has_dictionary);
+  ASSERT_GT(memory_usage2.memory, memory_usage1.memory);
+
+  columns.clear();
+  ParquetFileReader::MemoryUsage memory_usage_all
+      = reader_->EstimateMemoryUsage(false, columns, batch_size);
+  reader_->DebugPrint(ss, columns, batch_size, true);
+  ASSERT_GT(allocator_.MaxMemory(), memory_usage_all.memory);
+  ASSERT_TRUE(memory_usage_all.has_dictionary);
+  ASSERT_GT(memory_usage_all.memory, memory_usage2.memory);
```

Comment:
One way or another we need an internal API for constructing the Thrift metadata and creating the serialized file footer at least for testing purposes but also for writing Parquet files. The problem with ad hoc testing is that it doesn't protect you from bugs introduced by refactoring.

Without more comprehensive unit tests, this feature cannot be considered complete -- I am indifferent about whether it gets done in this JIRA or in a follow up JIRA, but I also don't want these tests to be thought of as "nice to have" rather than necessary.